### ISSUE TITLE
Adding a custom kernel transformer to the frontend server

### ISSUE THREAD

**Issue by Norbert515** [flutter/flutter#28523]
I'm working on an interactive widget editor and I need to write a custom kernel transformer.

The flutter/engine repository contains a frontend server which is a wrapper around the dart frontend server, in Flutter the compiled frontend server is included as `bin\cache\artifacts\engine\windows-x64\frontend_server.dart.snapshot`. It is used by the Flutter tool to generate the .dill files. 
What I'd like to do is to replace that frontend server with a custom one which adds a custom kernel transformer.

**What I tried:**
I downloaded the flutter engine with the gclient and synced it but when resolving the dependencies it wasn't able to find a few of them:
```yml
dependencies:
  args: any
  async: any
  charcode: any
  collection: any
  convert: any
  crypto: any
  front_end: any
  flutter_kernel_transformers: any
  kernel: any
  logging: any
  meta: any
  quiver: any
  package_config: any
  path: any
  source_span: any
  typed_data: any
  usage: any
  vm: any
```
For example, I was able to resolve the vm dependency by changing it to a path dependency and pointing to the downloaded vm package, but after changing a few things more and more errors kept popping up.

I tried implementing the changes using the dart repository directly but I ran into version conflicts and ultimately syncing dart + engine + flutter version didn't sound like a lot of fun.

I had problems building dart/ the engine on my machine (I'm on Windows).

**What worked:**
I've set up a Linux VM and downloaded the engine, I also was able to build it and therefore generate a `frontend_server.dart.snapshot`. Unfortunately, the build took around 4 hours.

Ideally, I'd like to open up the frontend_server as a separate package, add changes and compile the `frontend_server.dart.snapshot` using the correct version of dart/ the engine. 

I've seen @jacob314 implemented the track-widget-creation transformer, I'd love to hear how you went about it!

Any help is appreciated, thanks in advance.

**Comment by OsaXma** [comment]
I cannot exactly answer your question but you might find these docs helpful if you haven't looked at them yet:

[Setting Up The Framework Development Environment](https://github.com/flutter/flutter/wiki/Setting-up-the-Framework-development-environment)

You can also see in the right side panel:

**Engine repo**
- Setting up the Engine development environment
- Compiling the engine
- Debugging the engine
- ....
- ....
- ....
- etc

**Comment by Norbert515** [comment]
I was able to build the engine but it took a long time - ideally I'd like to only compile the frontend_server.

**Comment by jacob314** [comment]
Unfortunately you do need to rebuild the engine. I would also like it is that step could be avoided as it would make me more productive writing kernel transformers as well. Kernel transformers can easily have strange and surprising consequences in terms of app behavior and performance so we do need to be very careful about which transformers are run.
What are you looking for the kernel transformer to do? Potentially we can integrate the functionality into the transformer use by the Flutter Widget Inspector as well as features useful for one UI debugging tools are typically useful for all.

**Comment by Norbert515** [comment]
I've played around with the idea of having an IDE for Flutter written in Flutter for a while now. I released an early demo of a prototype a couple of weeks ago: 
- https://norbert515.github.io/widget_maker/website/
- https://medium.com/flutter-community/introducing-flutter-widget-maker-a-flutter-app-builder-written-in-flutter-231e8d959348

_What I need kernel transformers for:_

One core feature I thought of is visually editing widgets. You'd open up a .dart file and the IDE would add little run buttons beside the source code of widgets (just like IntelliJ does for main methods). When pressing that, the standalone widget would be rendered. From there on, you'd be able to drag&drop in widgets and live-reload changes. Every change would also be directly reflected in the source code.

The kernel transformer is needed to swap out widgets with modified counterparts and to add additional metadata for source code reconstruction (for example the source code of a function, because calling `toString()` doesn't print the source code).

I wrote down a bunch of thoughts in this readme _(keep in mind this is very experimental)_: 
https://gist.github.com/Norbert515/c68eacbf3acb4f8159d4cac546c6d96d

Having this capability in Flutter would definitely be awesome but I'd image it to quite hard to implement/ maintain it officially (I wouldn't be opposed if there is an intention to move in that direction though).
